### PR TITLE
feat(server): move authentication to tokens stored in the database

### PR DESCRIPTION
**Pull request by zackpollard**
This PR moves aways from JWTs to generated tokens that we store in the database. This has the following advantages:

- Tokens can be revoked at any time
- Tokens can have an infinite lifespan, no apps being logged out
- User can view a list of all tokens issued for their account
- API keys and user tokens now use SHA256 hashing for better performance

### CODE REVIEW

**Review by jrasm91 — APPROVED** [review]
:tada:

File: server/libs/domain/test/crypto.repository.mock.ts [hunk 1]
```diff
@@ -3,8 +3,9 @@ import { ICryptoRepository } from '../src';
 export const newCryptoRepositoryMock = (): jest.Mocked<ICryptoRepository> => {
   return {
     randomBytes: jest.fn().mockReturnValue(Buffer.from('random-bytes', 'utf8')),
-    compareSync: jest.fn().mockReturnValue(true),
-    hash: jest.fn().mockImplementation((input) => Promise.resolve(`${input} (hashed)`)),
+    compareBcrypt: jest.fn().mockReturnValue(true),
+    hashBcrypt: jest.fn().mockImplementation((input) => Promise.resolve(`${input} (hashed)`)),
+    hashSha256: jest.fn().mockImplementation((input) => `${input} (hashed)`),
     signJwt: jest.fn().mockReturnValue('signed-jwt'),
     verifyJwtAsync: jest.fn().mockResolvedValue({ userId: 'test', email: 'test' }),
```
Comment (jrasm91) [inline server/libs/domain/test/crypto.repository.mock.ts:10]: I think these aren't used any more, correct?

File: server/libs/infra/src/db/config/database.config.ts [hunk 2]
```diff
@@ -11,11 +11,11 @@ const baseDatabaseConfig: PostgresConnectionOptions = {
 };
 
 const envBasedDatabaseConfig = {
-  host: process.env.DB_HOSTNAME || 'immich_postgres',
```
Comment (alextran1502) [inline server/libs/infra/src/db/config/database.config.ts:14]: Is this related to development environment changes?

File: server/apps/immich/src/modules/immich-auth/strategies/user-auth-strategy.service.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+export const AUTH_COOKIE_STRATEGY = 'auth-cookie';
+
+@Injectable()
+export class UserAuthStrategy extends PassportStrategy(Strategy, AUTH_COOKIE_STRATEGY) {
+  constructor(private userService: UserService, private authService: AuthService) {
+    super();
+  }
+
+  async validate(request: Request): Promise<UserResponseDto> {
```
Comment (alextran1502) [inline server/apps/immich/src/modules/immich-auth/strategies/user-auth-strategy.service.ts:15]: @jrasm91 I think we need to return `AuthUserDto` here to have the same form with `api-key` and `public-share` strategy, corret?

Comment (jrasm91) [reply server/apps/immich/src/modules/immich-auth/strategies/user-auth-strategy.service.ts:15]: Yes, this should return an `AuthUserDto`, I believe it did before as well so we should keep that code/interface.

File: server/libs/infra/src/db/config/database.config.ts (hunk 2 above)
Comment (zackpollard) [reply server/libs/infra/src/db/config/database.config.ts:14]: With the default config like this, we can run typeorm commands from the host rather than within the server container, it couples with the nicer typeorm commands in package.json now. We always tell the user to set them in the .env file, so I think changing the default here is fine.

File: server/apps/immich/src/modules/immich-auth/strategies/user-auth-strategy.service.ts (hunk 3 above)
Comment (zackpollard) [reply server/apps/immich/src/modules/immich-auth/strategies/user-auth-strategy.service.ts:15]: Agreed, must've got changed in the refactor, change that now

File: server/libs/domain/src/auth/auth.service.ts [hunk 4]
```diff
@@ -112,49 +115,22 @@ export class AuthService {
...
-        if (payload?.userId && payload?.email) {
-          const user = await this.userCore.get(payload.userId);
-          if (user) {
-            return user;
-          }
-        }
-      }
-    } catch (e) {
-      return null;
+  public async validate(headers: IncomingHttpHeaders): Promise<UserResponseDto> {
```
Comment (alextran1502) [inline server/libs/domain/src/auth/auth.service.ts:118]: This is still returning `UserResponseDto` instead of `AuthUserDto`. Should we explicitly convert to `AuthUserDto`?
